Enable minimal SteamAPI integration for usage time tracking (editor only).

bruvzg opened the pull request:
Disabled by default, can be enabled with (`steamapi=yes` build option), ~`steam_appid.txt`~ (it's only needed when running executable outside of Steam download) and `steam_api` library from the SDK should be added to distribution for this feature to work.

It initializes Steam API to inform Steam that Godot editor is running, but not using it in any way.

Fixes #18233

bruvzg committed:
Enable optional minimal SteamAPI integration for usage time tracking (editor only).

akien-mga commented on main/steam_tracker.h:
```diff
@@ -0,0 +1,53 @@
+typedef bool (*SteamAPI_InitFunction)();
+typedef void (*SteamAPI_ShutdownFunction)();
+
+class SteamTracker {
```
Would be good to add some docs to clarify what this is used for.

bruvzg replied:
Added some info and likes to SteamAPI documentation.

akien-mga commented on main/SCsub:
```diff
@@ -10,6 +10,9 @@ env_main = env.Clone()
 
 env_main.add_source_files(env.main_sources, "*.cpp")
 
+if env["steamapi"]:
```
Aside from being only needed for the editor's time tracking currently, the code itself isn't really editor-specific, so you could check only `STEAMAPI_ENABLED` in the code and here do:
```suggestion
if env["steamapi"] and env["tools"]:
```

Calinou commented:
I wonder if this will make Remote Play Together usable with the editor on Steam. This would allow for limited pair programming when using Steam (although the running project will only be visible by one of the users).

bruvzg commented:
> I wonder if this will make Remote Play Together usable with the editor on Steam.

Seems like implementing "invite to play" is optional, and the rest of [Remote Play API](https://partner.steamgames.com/doc/api/ISteamRemotePlay) is for reading information about current session only. It might be possible that enabling it in the Steam admin panel is enough, or maybe it will just work without any changes.

> > "Users can invite friends from their friends list in the Steam overlay, and if you want, you can create UI to invite friends directly from your game using the ISteamRemotePlay::BSendRemotePlayTogetherInvite() function."
>
> > This feature is enabled automatically for games that are listed with Local Multiplayer, Local Co-op, or Shared/Split Screen capability, but you can enable or disable it yourself to match whether this feature is appropriate for your game.
> 
> From https://partner.steamgames.com/doc/features/remoteplay

But I'm not sure how it's going to handle input.

bruvzg commented:
With the Godot executable in the Steam distro substituted with one from this PR, it is usable with Steam Link, it's not loosing connection after opening a project. So it probably will be somewhat usable with Remote Play Together as well.

akien-mga commented:
CC @Gramps - I don't think this would affect https://github.com/Gramps/GodotSteam but it's worth checking.

bruvzg commented:
> I don't think this would affect https://github.com/Gramps/GodotSteam but it's worth checking.

It should not affect exported projects, but it might affect projects started from the editor (might need an extra check for editor/project manager flag to avoid this).

Gramps commented:
Without actually testing it with both this and GodotSteam enabled, I'm not sure if there would be any conflicts.  This only seems to initialize and shut down Steam.  What happens if this is run outside of Steam with the steam_appid.txt missing?

I've never had a problem with Steam tracking my time when using the editor run separately, but it does track it to the game I'm working on.

bruvzg commented:
> What happens if this is run outside of Steam with the steam_appid.txt missing?

`SteamAPI_Init` prints an error to the console and returns `false`.

Gramps commented:
Initialization fails but it still tracks the time?  Interesting...

bruvzg commented:
> Initialization fails but it still tracks the time? Interesting...

No, it won't track the time without `steam_appid.txt` or running from the Steam.

akien-mga commented:
> > What happens if this is run outside of Steam with the steam_appid.txt missing?
> 
> `SteamAPI_Init` prints an error to the console and returns `false`.

Only if the Steam library is included, right? As I use the same binaries for Steam, itch and the Godot repositories, so for non-Steam users I expect it not to print anything Steam related.

bruvzg commented:
> Only if the Steam library is included, right?

If library is not icluded, `open_dynamic_library` won't be able to open it, but have it's own error message.

I'll add some extra library existance checks.

Gramps commented:
I'm still playing catch-up on the original issue.  From what I understand this is fix people launching the editor through Steam and when it opens a project it stops tracking time and only on Windows?

bruvzg commented:
> I'm still playing catch-up on the original issue. From what I understand this is fix people launching the editor through Steam and when it opens a project it stops tracking time and only on Windows?

Yes, on macOS it's also the same.

Gramps commented:
Gotcha.  I'm about to start compiling it along with GodotSteam to see if they play nice.  Will report back in a bit.

bruvzg commented:
Added a librarby existance cheks, so it won't print any errors if it's missing.
Also added `project_manager || editor` check to prevent it from tracking running projects.

Gramps commented:
The compiled version with this and GodotSteam seems to work fine.  I didn't notice any issues in console and the GodotSteam functions worked as expected.  That being said, I don't see any reason why someone would run both together but it's good they don't conflict just in case someone did.

akien-mga commented on main/steam_tracker.cpp:
```diff
@@ -0,0 +1,102 @@
+	if (OS::get_singleton()->has_feature("linuxbsd")) {
+		path = OS::get_singleton()->get_executable_path().get_base_dir().plus_file("libsteam_api.so");
+		if (!FileAccess::exists(path)) {
+			path = OS::get_singleton()->get_executable_path().get_base_dir().plus_file("../lib").plus_file("libsteam_api.so");
```
Would that work to load the library without setting a `RPATH`? Does Steam automatically add it to `LD_LIBRARY_PATH`?

Gramps replied:
Probably?  I know Steam on Linux has some weird issues sometimes.  We usually tell people to put the .so file in the directory with the editor but even then some need a launcher script to link in that file.  Some people reported they needed the .so file in the project's actual directory (not the editor).  It can vary by distro and, I believe, how Steam it actually installed; where it puts its library files and what versions.

Infinixius commented:
Any progress on this?

Gramps commented:
Very cool work!

akien-mga commented:
This looks good to me, but before merging I wanted to give this a proper try on Steam for Windows. It's a bit late now for 4.0, I'll postpone to 4.1, and possibly get it tested via 3.6 betas first.

caimantilla commented:
Is there any chance of this being included in a future 4.1 release?

bruvzg commented:
GitHub seems to completely ignore force pushes to the branches with "invalid characters", it was impossible to update original PR, so reopened it as https://github.com/godotengine/godot/pull/79126

YuriSizov commented:
> Is there any chance of this being included in a future 4.1 release?

Not in 4.1.x, it would be out of scope. But it can be added to 4.2.
